### PR TITLE
improvements for resolving java home

### PR DESCRIPTION
**Pull request by kasiaMarek**
connected to https://github.com/scalameta/metals-vscode/issues/1464
resolves: https://github.com/scalameta/metals/issues/6168

### CODE REVIEW

File: packages/metals-languageclient/src/getJavaHome.ts [hunk 1]
```diff
@@ -26,23 +27,26 @@ export type JavaVersion = "11" | "17" | "21";
...
 ): Promise<boolean> {
-  const javaBins = [
-    path.join(javaHome, "bin", "java"),
+  const allPossibleJavaBins = [
     path.join(javaHome, "bin", "java.exe"),
-    path.join(javaHome, "bin", "jre", "java"),
     path.join(javaHome, "bin", "jre", "java.exe"),
-  ].filter(fs.existsSync);
+    path.join(javaHome, "bin", "java"),
+    path.join(javaHome, "bin", "jre", "java"),
```
Comment (tgodzik) [inline packages/metals-languageclient/src/getJavaHome.ts:47]: Actually, I think jre might be problematic and will not work for us.

File: packages/metals-languageclient/src/getJavaHome.ts [hunk 2]
```diff
@@ -26,23 +27,26 @@ export type JavaVersion = "11" | "17" | "21";
  * @param javaVersion metals.javaVersion value as read from the configuration or default
  */
 export async function getJavaHome(
-  javaVersion: JavaVersion
+  javaVersion: JavaVersion,
+  outputChannel: OutputChannel
 ): Promise<string | undefined> {
-  const fromEnvValue = await fromEnv(javaVersion);
+  const fromEnvValue = await fromEnv(javaVersion, outputChannel);
```
Comment (tgodzik) [inline packages/metals-languageclient/src/getJavaHome.ts:33]: Any chance to forward coursier downloading information?

Comment (kasiaMarek) [reply packages/metals-languageclient/src/getJavaHome.ts:33]: We are forwarding courier downloading info.

File: packages/metals-languageclient/src/getJavaHome.ts [hunk 3]
```diff
@@ -26,44 +27,45 @@ export type JavaVersion = "11" | "17" | "21";
...
-      encoding: "utf8",
-    });
-    const javaInfoStr = javaVersionOut.stderr as string;
-    const matches = javaInfoStr.match(versionRegex);
-    if (matches) {
-      return matches[0].slice(0, 2) == javaVersion;
-    }
+  const javaBin = path.join(javaHome, "bin", "java");
+  const javaVersionOut = await spawn(javaBin, ["-version"], {
+    encoding: "utf8",
```
Comment (kasiaMarek) [inline packages/metals-languageclient/src/getJavaHome.ts:45]: Actually, let's wrap this in a try.

**Review by tgodzik — APPROVED** [review]

**Pull request merged**